File: Homeworks/Chapter1/Solutions/Soal5.cpp
#include  <iostream>
using namespace std;
int main(){

	int t;
	cin >> t;
	while(t--){

		int n,m,f=1;
		cin >> n >> m;
		
		int **mat = new int * [n];
		for(int i=0;i<n;i++)
			mat[i] = new int [m];
		for(int i=0;i<n;i++){
			for(int j=0;j<m;j++){
				cin >> mat[i][j];
			}
		}
		for(int i=0;i<n;i++){
			int cwd[11]={0};
			for(int j=0;j<m-1;j++){
				int diff = abs(mat[i][j]-mat[i][j+1]);
				if(diff>m-1||diff==0){
					continue;
				}
				cwd[diff]++;
			}
			for(int c=1;c<=m-1;c++){
				if(cwd[c]!=1){
					f=0;
				}
			}
			if(!f)break;

		}
		for(int j=0;j<m;j++){
			int cwd[11]={0};
			for(int i=0;i<n-1;i++){
				int diff = abs(mat[i][j]-mat[i+1][j]);
				if(diff>n-1||diff==0){
					continue;
				}
				cwd[diff]++;
			}
			for(int c=1;c<=n-1;c++){
				if(cwd[c]!=1){
					f=0;
				}
			}
			if(!f)break;
		}
		for(int i=0;i<n;i++)
			delete[] mat[i];
		delete[] mat;
		if(!f){
			cout<<"Wicked"<<endl;
			continue;
		}
		cout<<"Tame"<<endl;
	}
	return 0;
}

File: Homeworks/Chapter1/Solutions/Soal5.java
package DS_Chapter1 ;

import java.util.Scanner;

public class Soal5 {
    static Scanner in = new Scanner(System.in);
    public static void main(String[] args){
        int numberOfMatrices = in.nextInt();
        for (int i = 0; i<numberOfMatrices; i++){
            String status = "Tame";
            int rows = in.nextInt();
            int columns = in.nextInt();
            int[][] matrix = new int[rows][columns];
            for (int r = 0; r<rows; r++){
                int[] row = new int[columns];
                for (int c = 0; c<columns; c++){
                    row[c] = in.nextInt();
                    matrix[r][c] = row[c];
                }
                if (!isTame(row)){
                    status = "Wicked";
                    break;
                }
            }
            if (status != "Wicked"){
                for (int c = 0; c<columns; c++){
                    int[] column = new int[rows];
                    for (int r = 0; r<rows; r++){
                        column[r] = matrix[r][c];
                    }
                    if (!isTame(column)){
                        status = "Wicked";
                        break;
                    }
                }
                System.out.println(status);
            } else {
                System.out.println(status);
            }
        }

    }

    public static boolean isTame(int... arr){
        boolean b = true;
        int n = arr.length;
        int [] khosh = new int[n-1];
        for (int i = 0; i < n-1; i++)
            khosh[i] = Math.abs(arr[i] - arr[i+1]);
        for (int i = 0; i < n-2; i++) {
            if(Math.abs(khosh[i] - khosh[i+1]) != 1)
                b = false;
        }
        return b;
    }
}